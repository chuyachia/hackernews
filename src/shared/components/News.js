import React from "react";
import axios from "axios";
import styled , { keyframes } from 'styled-components';
import Comments from "./Comments";

const UnstyledUl = styled.ul`
    list-style:none;
    margin: 0;
    padding: 0;
    display:${props => props.visible ? 'initial' : 'none'};
    width:100%;
`;

const Newsitem =styled.li`
    display:flex;
    flex-flow: row wrap;
    & * {
        flex: 1 1 100%;
        word-wrap:break-word;
        white-space: pre-wrap;
    }
    &:nth-child(even) {
        background-color:#d8e9f3
    }
    &:nth-child(odd) {
        background-color:#b2d3e6
    }
    padding:0.5rem;
`;

const Author = styled.span`
    color:grey;
    @media all and (min-width: 800px){
        flex: 1 1 auto; 
    }
`;

const ShowHide= styled.span`
    @media all and (min-width: 800px){
        flex: 0 1 auto;
    };
    text-align:right;
    cursor:pointer;
`;

const Title = styled.a`
    font-family: 'Montserrat';
    color: inherit;
    text-decoration: inherit;
    @media all and (min-width: 800px){
        flex: 0 1 auto;
    }
`;

const rotate = keyframes`
    0% { transform: rotate(0deg); }
    100% { transform: rotate(360deg); }
`;

const Loader = styled.div`
    border: 16px solid lightgrey;
    border-top: 16px solid grey; 
    border-radius: 50%;
    width: 50px;
    height: 50px;
    margin: 0 auto;
    animation: ${rotate} 2s linear infinite;
`;





const hnUrl= "https://hacker-news.firebaseio.com/v0/";
class News extends React.Component {
    constructor(){
        super();
        this.state={
            comments:null,
            showcomments:false
        };
    }
    showComments(){
        if (!this.state.comments&&this.props.data.kids){
            var promises = this.props.data.kids.map(function(id){
                return axios.get(hnUrl+'item/'+id+'.json');
            });
            axios.all(promises)
            .then((results)=>{
                results = results.filter(r=>!r.data.deleted);
                results = results.slice(0,21);
                this.setState({
                    comments:results.map(function(c,i){return <Comments key={i} data={c.data}/>})
                });
            });
        }
        this.setState({showcomments:this.state.showcomments?false:true});
        
    }
    render(){
        return(
            <Newsitem>
                {<Title href={this.props.data.url} target="_blank"  title="Open link">{this.props.data.title}&nbsp;</Title>}
                {<Author>by&nbsp;{this.props.data.by}</Author>}
                {<ShowHide id="showhide" onClick={()=>this.showComments()}>
                {this.state.showcomments?<i title="Hide comments" class="fas fa-sort-up"></i>
                :<i title="Show comments" class="fas fa-sort-down"></i>}</ShowHide>}
                {<UnstyledUl visible={this.state.showcomments} onClick={()=>this.showComments()}>
                        {this.props.data.kids?(this.state.comments?this.state.comments:<Loader/>):<li>No comment to show...</li>}
                </UnstyledUl>}
            </Newsitem>
            );
    }
}

export default News;