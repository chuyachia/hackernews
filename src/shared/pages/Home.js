import React from "react";
import axios from "axios";
import styled from 'styled-components';
import News from '../components/News';

const Container = styled.div`
    display:flex;
    flex-flow: row wrap;
    & > * {
    width:100%;
     flex: 1 1 auto;
    }
`;

const Title=styled.div`
    text-align:center;
    & > i {
      color:grey;  
    }
`;



const Newslist = styled.ul`
    list-style:none;
    padding-left:0;
    @media all and (min-width: 800px){
        padding-left:10%;
        padding-right:10%;
    }
`;

const Footer=styled.div`
    text-align:center;
    & > a {
        color: inherit;
        text-decoration: underline dashed;
        text-underline-position: under;
    }
`
const hnUrl= "https://hacker-news.firebaseio.com/v0/";
class Home extends React.Component {
    constructor(){
        super();
        this.state={
            news:[]
        };
    }
    componentDidMount(){
        axios.get(hnUrl+'topstories.json')
        .then((results)=>{
            var promises = results.data.slice(0,10).map(function(id){
                return axios.get(hnUrl+'item/'+id+'.json');
            });
            return  axios.all(promises);
        }).then((results)=>{
            this.setState({
                news:results.map(function(r){return r.data})
            });
        })
        .catch(function(err){
            console.log(err);
        });
    }
    render(){
        return(
         <Container>
             <Title>
                 <h3>Top 10 Hacker News Stories</h3>
                 <i>With top 20 comments and their replies</i>
             </Title>
             <Newslist>
             {
             this.state.news.map((n,i)=>(<News key={i} data ={n}/>))
             }
             </Newslist>
             <Footer>Made by <a href="https://github.com/chuyachia" target="_blank">Chu-Ya Chia</a></Footer>
        </Container>);
    }
}


export default Home;
