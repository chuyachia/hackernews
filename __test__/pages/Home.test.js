import { shallow } from 'enzyme';
import React from 'react';
import Home from '../../src/shared/pages/Home';
import News from '../../src/shared/components/News';
import * as enzyme from 'enzyme';
import ReactSixteenAdapter  from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

enzyme.configure({ adapter: new ReactSixteenAdapter() });


const flushPromises = () => new Promise(resolve => setImmediate(resolve));

describe('Home page test',()=>{
      let wrapper;
      beforeEach(async ()=>{
        wrapper = shallow(<Home />);
        await flushPromises();
        wrapper.update();
      })
      it('Home page snapshot',() => {
        expect(toJson(wrapper)).toMatchSnapshot();
      });
      it('Home page has 10 news components',() => {
        expect(wrapper.find(News)).toHaveLength(10);
      });
})
