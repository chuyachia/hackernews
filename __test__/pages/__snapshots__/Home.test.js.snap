// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page test Home page snapshot 1`] = `
<styled.div>
  <styled.div>
    <h3>
      Top 10 Hacker News Stories
    </h3>
    <i>
      With top 20 comments and their replies
    </i>
  </styled.div>
  <styled.ul>
    <News
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="0"
    />
    <News
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="1"
    />
    <News
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="2"
    />
    <News
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="3"
    />
    <News
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="4"
    />
    <News
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="5"
    />
    <News
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="6"
    />
    <News
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="7"
    />
    <News
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="8"
    />
    <News
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="9"
    />
  </styled.ul>
  <styled.div>
    Made by 
    <a
      href="https://github.com/chuyachia"
      target="_blank"
    >
      Chu-Ya Chia
    </a>
  </styled.div>
</styled.div>
`;
